### ISSUE TITLE
Create automation pallet for EuroBrasil project

### ISSUE THREAD

**Issue by ebma** [pendulum-chain/pendulum#357]
## Context

The context can be found in [this](https://www.notion.so/satoshipay/Automating-BRZ-EUR-on-Pendulum-4037d39f0b434a5ab2e299195b228b33) Notion page.

## Pallet configuration

**Config Items**

- `PalletId` - needed to derive the pallets on-chain account (holding BRZ and EURC)

**Storage Items**

- `NablaRouterAddress` - the address of Nabla’s `router` smart contract
- `EurcErc20Address` - the address of the ERC-20 wrapper contract for the EURC token
- `BrzErc20Address` - the address of the ERC-20 wrapper contract for the BRZ token

- `OfframpAccount` - the address of the designated off-ramping account (chosen by Mykobo)

**Extrinsics**

- `setNablaRouterAddress(address)` - sets the address of the `router` contract
- `setErc20Adresses(eurcAddress, brzAddress)` - sets the addresses of the erc-20 contracts
- `setOfframpAccount(address)` - sets the address of the offramping account

**Functions (not exposed as extrinsic but available to other pallets)**

- `processEurcOfframp(amount, reference)` - this function is called by the XCM config of our runtime. It handles the conversion from BRZ → EUR and transfers the tokens to the offramping account. `amount` is the amount of BRZ tokens that are to be offramped and `reference` is the reference for Mykobo that was encoded by the `MultiLocation`

**Coupling to other pallets**

- *Contracts* - needed to call the nabla contracts
- *Tokens* - needed for the transfer
- *System* - needed to set the remark

### Prerequisites
Upon receiving an XCM transfer of BRZ from Moonbeam targeting our automation pallet, the XCM config deposits the respective amount of BRZ into this pallet’s on-chain account and calls the `processEurcOfframp()` function, supplying the processed amount of BRZ and the reference number required by Mykobo (which is encoded in the `MultiLocation`).

### Implementation of the `processEurcOfframp()` function

The `processEurcOfframp()` function uses the `call()` function of the *Contracts* pallet to call the `swapExactTokensForTokens()` function on Nabla’s `router` smart contract to swap the `amount` of BRZ tokens to EURC tokens. As a result, it will receive the corresponding amount of EURC on its on-chain account. 
The `swapExactTokensForTokens()` function needs to be called using the addresses of the ERC-20 contracts. The signature of the function can be found [here](https://github.com/0xamberhq/contracts/blob/a4610d4b3431e1b00dacf6449d2dd6f85f48a48f/contracts/src/core/Router.sol#L85-L91).

Next, the pallet calls the `remark()` function of the System pallet with the supplied `reference` number and the `transfer()` function of the Tokens pallet to transfer the swapped EURC to the offramping account.

**Comment by ebma** [comment]
@pendulum-chain/product this ticket is required for the EuroBrasil project. It's about adding the pallet that automates the currency conversion and offramping. Please assign a high priority to it.

**Comment by prayagd** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Pendulum-6408a0365cda59c57d39bee5/issues/pendulum-chain/pendulum/357?planning-poker) with Zenhub @b-yap @ebma @gianfra-t @TorstenStueber @bogdanS98

**Comment by ebma** [comment]
Thanks for the comment @TorstenStueber. I favor the former of the approaches you mentioned and I changed the description of this ticket as to how I imagine this to work.
I introduced an extra account for the buffer pool. In theory, we don't even need to have two accounts and we could just have one pallet account that we charge with some EURC tokens which would then handle the buffering, but I think it's cleaner to separate these two concerns

**Comment by TorstenStueber** [comment]
Yeah, I thought about it and I think it is okay and natural that we control this "buffer account". At the end Pendulum as a blockchain is the neutral system and then Pendulum Pay is a service built on top of that that we provide (whereas "we" is in that sense a separate entity, e.g., SatoshiPay) and in order to provide a smooth service we control this account.

A few more thoughts I have:
* What is the potential attack vector someone has if they get access to one of the authorized accounts? They could call `addEurAmountForReference` and empty the buffer account this way. However, this would not be freely transferable but would go through a offramping with Mykobo for a KYC'd account so that I think that the danger is limited.

* What happens if the buffer account is not covered sufficiently (let's say some attack on Nabla leads to heavy slippage and therefore the exchange rate is much worse than expected). We would need to define what happens in this situation. For now we could of course just not complete the transaction and leave the funds on the pallet account. That might be enough for the MVP (what do you say @vadaynujra?)

* We need to issue the `addEurAmountForReference` transaction as soon as we display the QR code to the user. Once the funds are onramped we would expect that there is already an entry in the `ReferenceToEurAmount`. The timing here will most likely work but we again need to define what happens if such an entry is not yet in the storage. Again, for the MVP it could be enough to just not complete the transaction.

* Usually the offramp would just be a batch transaction consisting of a transfer and a remark. This way the indexer can easily report that a certain transfer is associated with a specific reference. I don't think that this would work here. The `remark` intrinsic [is actually not doing anything](https://github.com/paritytech/polkadot-sdk/blob/1f2ccaea05c8508f5cc916f8ae0ba60c102c5cf3/substrate/frame/system/src/lib.rs#L472C33-L472C33) and the whole purpose is that by calling this extrinsic as a transaction, the `remark` would just be stored in the transaction set of the block. In our case the `remark` would not be stored at all. So I think we need to find a different solution.

**Comment by gianfra-t** [comment]
For the remark issue, it is true that it does not really perform any state change. But as long as the batch call from this pallet is registered as it would be with any extrinsic, then the indexer should pick it up and append the data.

I guess the true question is if it indeed gets registered as a call. Is this what you are referring to @TorstenStueber?

**Comment by ebma** [comment]
> We could introduce a new event in the automation pallet that contains all the important pieces of information and just emit that event. Then our indexer would allow to watch these events.

Right, I guess this is the best option we have.

> shall we continue with the offramp if the buffering does not work as expected?

Regarding this, I think if we don't want to continue without buffering and thus not achieving the designated exchange rate, we could alternatively stop processing the request in the pallet and wait until the buffer pool is charged again. We could add an extra extrinsic like `continueOfframpForReference(reference)` callable by anyone, which would retry the processing of the offramp requests, re-checking if the buffer pool account holds enough tokens this time. This approach means that the execution of offramp requests does not always happen immediately and we'd need to add some kind of monitoring, but at least we could then again guarantee that the targeted exchange rate is achieved.

**Comment by gianfra-t** [comment]
> We could introduce a new event in the automation pallet that contains all the important pieces of information and just emit that event. Then our indexer would allow to watch these events.

I think this is better, we are already creating a new pallet so it is not so much extra modifications. I also cannot think of how we could do this with the `remark_with_event` since we won't have any identification of the transaction internally.

**Comment by TorstenStueber** [comment]
Okay, then let's add the new event.

For @ebma's point (introduce `continueOfframpForReference(reference)`): I think this makes a lot of sense. For every swap on Nabla we can state what minimum amount we want to receive in return. This is what we should use to ensure that the exchange rate is not too bad. Then there are two "breaking points":
* Nabla cannot give us the minimum exchange rate we need or
* Even if the Nabla swap works there are not enough funds in the buffer account to pay the difference.

If this is not too complex to implement, then I think it is worthy to do this already for the MVP. Hard to tell as we need to be very careful not to introduce vulnerabilities. Let's split it into an extra ticket, though and then estimate it separately.

**Comment by ebma** [comment]
I updated the description. It now mentions two events and the 'continuation' of paused transfers is covered in #378.

**Comment by prayagd** [comment]
@ebma Thank you, expecting this is ready to work moving it to the ready column

**Comment by ebma** [comment]
I changed the description slightly. I think we should still take the assumption that we will have one reference number that uniquely identifies each forex transfer and a designated off-ramp account on-chain. But we'll only know once we find another issuer for the EUR asset.

**Comment by TorstenStueber** [comment]
@annatekl @prayagd I think that we can move this back to Ready now as @ebma adapted the description.
